code/modules/atmospherics/machinery/components/trinary_devices/hypertorus.dm

```diff
@@ -0,0 +1,422 @@
+#define PARTICLE_CHANCE_CONSTANT 			(-20000000)
+
+/obj/machinery/atmospherics/components/unary/hypertorus
+	icon = 'icons/obj/atmospherics/components/hypertorus.dmi'
+	icon_state = "core"
+
+	name = "thermomachine"
+	desc = "Heats or cools gas in connected pipes."
+	anchored = TRUE
+	density = TRUE
+	circuit = /obj/item/circuitboard/machine/thermomachine
+	var/icon_state_open = "moderator_input"
+	var/icon_state_off = "moderator_input"
+	var/active = FALSE
+
+/obj/machinery/atmospherics/components/unary/hypertorus/Initialize()
+	. = ..()
+	initialize_directions = dir
+
+/obj/machinery/atmospherics/components/unary/hypertorus/attackby(obj/item/I, mob/user, params)
+	if(!on)
+		if(default_deconstruction_screwdriver(user, icon_state_open, icon_state_off, I))
+			return
+	if(default_change_direction_wrench(user, I))
+		return
+	if(default_deconstruction_crowbar(I))
+		return
+	return ..()
+
+/obj/machinery/atmospherics/components/unary/hypertorus/default_change_direction_wrench(mob/user, obj/item/I)
+	if(!..())
+		return FALSE
+	if(!anchored)
+		return FALSE
+	SetInitDirections()
+	var/obj/machinery/atmospherics/node = nodes[1]
+	if(node)
+		if(src in node.nodes) //Only if it's actually connected. On-pipe version would is one-sided.
+			node.disconnect(src)
+		nodes[1] = null
+	if(parents[1])
+		nullifyPipenet(parents[1])
+
+	atmosinit()
+	node = nodes[1]
+	if(node)
+		node.atmosinit()
+		node.addMember(src)
+	SSair.add_to_rebuild_queue(src)
+	return TRUE
+
+/obj/machinery/atmospherics/components/unary/hypertorus/fuel_input
+	name = "fuel_input"
+	desc = "fuel_input"
+	icon_state = "fuel_input"
+	icon_state_open = "fuel_input"
+	icon_state_off = "fuel_input"
+	circuit = /obj/item/circuitboard/machine/hypertorus/fuel_input
+
+/obj/machinery/atmospherics/components/unary/hypertorus/waste_output
+	name = "waste_output"
+	desc = "waste_output"
+	icon_state = "waste_output"
+	icon_state_open = "waste_output"
+	icon_state_off = "waste_output"
+	circuit = /obj/item/circuitboard/machine/hypertorus/waste_output
+
+/obj/machinery/atmospherics/components/unary/hypertorus/moderator_input
+	name = "moderator_input"
+	desc = "moderator_input"
+	icon_state = "moderator_input"
+	icon_state_open = "moderator_input"
+	icon_state_off = "moderator_input"
+	circuit = /obj/item/circuitboard/machine/hypertorus/moderator_input
+
+/obj/machinery/hypertorus
+	name = "hypertorus_core"
+	desc = "hypertorus_core"
+	icon = 'icons/obj/atmospherics/components/hypertorus.dmi'
+	icon_state = "core"
+	move_resist = INFINITY
+	anchored = FALSE
+	density = TRUE
+	circuit = /obj/item/circuitboard/machine/thermomachine
+	var/active = FALSE
+
+/obj/machinery/hypertorus/ComponentInitialize()
+	. = ..()
+	AddComponent(/datum/component/simple_rotation,ROTATION_ALTCLICK | ROTATION_CLOCKWISE | ROTATION_COUNTERCLOCKWISE | ROTATION_VERBS )
+
+/obj/machinery/hypertorus/wrench_act(mob/user, obj/item/I)
+	. = ..()
+	if(!active)
+		anchored = !anchored
+	else
+		message_admins("Is active")
```

Comment:
Debug lines.

Reply:
Yeah and I'll need to add many many more of them